### PR TITLE
Task manager extension

### PR DESCRIPTION
**Pull request by Deexie**
The PR adds changes to task manager that allow more convenient integration with modules.  

Introduced changes:
- adds internal flag in task::impl that allows user to filter too specific tasks
- renames `parent_data` to more appropriate name `task_info`
- creates `tasks/types.hh` which allows using some types connected with task manager without the necessity to include whole task manager
- adds more flexible version of `make_task` method

### CODE REVIEW

**Comment by Deexie** [comment]
- added a commit that updates ttl on each shard

File: api/api-doc/task_manager.json [hunk 1]
```diff
@@ -49,6 +49,14 @@
                         "type":"string",
                         "paramType":"path"
                     },
+                    {
+                        "name":"internal",
+                        "description":"Boolean flag indicating whether the internal tasks should be shown (false on default)",
```
Comment (bhalevy) [inline api/api-doc/task_manager.json:54]: nit: s/the internal/internal/
s/on default/by default/

File: tasks/types.hh [hunk 2]
```diff
@@ -14,4 +14,16 @@ namespace tasks {
 
 using task_id = utils::tagged_uuid<struct task_id_tag>;
 
+struct task_info {
+    task_id id;
+    unsigned shard;
+
+    task_info() noexcept : id(task_id::create_null_id()) {}
+    task_info(task_id parent_id, unsigned parent_shard) noexcept : id(parent_id), shard(parent_shard) {}
```
Comment (bhalevy) [inline tasks/types.hh:22]: now that the class is not necessarily about parent, maybe the constructor params should not mention parent, as I guess this would be part of the context of the constructed object.

File: tasks/task_manager.hh [hunk 3]
```diff
@@ -344,12 +354,22 @@ public:
...
             if (parent_d) {
-                co_await _tm.container().invoke_on(parent.get_owner_shard(), [task = std::move(parent), child = std::move(task)] (task_manager& tm) mutable {
-                    task->add_child(std::move(child));
-                });
+                co_await add_child(std::move(parent), std::move(task));
             }
             co_return id;
         }
+
+        // Responsibility for task_impl_ptr correctness is put on caller.
```
Comment (bhalevy) [inline tasks/task_manager.hh:362]: I don't understand what does that responsibility for correctness mean.
Please be more specific about what the caller is responsible for.

File: tasks/task_manager.hh [hunk 4]
```diff
@@ -344,12 +354,22 @@ public:
...
+        future<task_ptr> make_task(task::task_impl_ptr task_impl_ptr, task_info parent_d = task_info{}) {
+            auto task = make_lw_shared<task_manager::task>(std::move(task_impl_ptr));
+            foreign_task_ptr parent;
+            if (parent_d) {
+                parent = co_await get_parent(parent_d);
+                task->get_status().sequence_number = parent->get_sequence_number();
+                co_await add_child(std::move(parent), make_foreign(task));
```
Comment (bhalevy) [inline tasks/task_manager.hh:369]: this is not very efficient since we submit_to the parent shard both in get_parent and in add_child.
if we open coded the two we could just get the parent sequence number when looking it up and update the child while at it, and eventually update the child sequnece_number, requiring only a single submit_to.
Although performance is not critical on this path, we shouldn't waste resource.

With that, the existing make_task should probably create the task_impl_ptr before looking up the parent, and then call this function to tie them together.

File: tasks/task_manager.hh [hunk 5]
```diff
@@ -415,8 +415,10 @@ protected:
     }
 private:
     future<> update_task_ttl() noexcept {
-        _task_ttl = _cfg.task_ttl.get();
-        return make_ready_future<>();
+        auto new_ttl = _cfg.task_ttl.get();
+        return container().invoke_on_all([new_ttl] (auto& tm) {
+            tm._task_ttl = new_ttl;
+        });
```
Comment (bhalevy) [inline tasks/task_manager.hh:421]: If this is fixing an existing bug, please open an issue about it, and probably submit it in a separate PR so it can be expedited.

**Review by bhalevy — CHANGES_REQUESTED** [review]

**Review by bhalevy — APPROVED** [review]
lgtm
